### ISSUE TITLE
New ColorBar Converter

### ISSUE THREAD

**Issue by BobSimons** [ERDDAP/erddap#44]
ERDDAP has a group of Converters which do useful things (usually conversions) related to ERDDAP. See
[https://coastwatch.pfeg.noaa.gov/erddap/convert/index.html](https://coastwatch.pfeg.noaa.gov/erddap/convert/index.html)
It would be nice to add a ColorBar converter which takes a color bar specification (e.g., palette=Rainbow, Range=1 to 10000, Scale=log, NSections=4, Continuity=Continuous) and then converted a data value into a color or vice versa.

Skills required: Java programming.

Difficulty: Pretty easy. There are already Converters which could be used as templates for this converter. There is code in ERDDAP to convert a data value to a color. You probably have to write code for the reverse, but that is probably straightforward. Maybe 1-2 weeks of work (after you figure out ERDDAP).

Mentor: Bob Simons (main author of ERDDAP)

Please also read the Programmer's Guide at [https://coastwatch.pfeg.noaa.gov/erddap/download/setup.html#programmersGuide](https://coastwatch.pfeg.noaa.gov/erddap/download/setup.html#programmersGuide)
especially the "Judging Your Code Contributions" section.

This was requested by Lynne Tablewski and others.

**Comment by GAyan17** [comment]
Hello, This issue looks good to me can I work on it? Though I'm still a begginner in terms of JAVA Development but wanting to learn as I go, would be a great help if you guide me in some way to how to start working for this issue.

**Comment by BobSimons** [comment]
[I'm sorry for being so slow to answer (long weekend and mini emergency at work yesterday).]

Excellent! Yes, you can work on it. Here are suggested steps:

1. Read [the Programmer's Guide](https://coastwatch.pfeg.noaa.gov/erddap/download/setup.html#programmersGuide)
2.  Use ERDDAP a little so you understand what it is and how it works. [https://coastwatch.pfeg.noaa.gov/erddap/index.html](https://coastwatch.pfeg.noaa.gov/erddap/index.html)
3. Specifically, play around with a Make A Graph web page, so you understand color bars in ERDDAP and their options [https://coastwatch.pfeg.noaa.gov/erddap/griddap/erdMWsstdmday_LonPM180.graph](https://coastwatch.pfeg.noaa.gov/erddap/griddap/erdMWsstdmday_LonPM180.graph)
4. Play around with the converters so you understand what they are and how they work (one web page with instructions, a form, Details, and a description of how to do the conversion via the underlying service) [https://coastwatch.pfeg.noaa.gov/erddap/convert/index.html](https://coastwatch.pfeg.noaa.gov/erddap/convert/index.html)
5. Think in your head what your converter is going to look like. Email me to make sure we agree.
6. Get ERDDAP set up on your computer, so you can compile it, run ERDDAP on your computer, and later test your code. Being able to install Tomcat and an application is a very useful skill to have since they are both widely used.  [https://coastwatch.pfeg.noaa.gov/erddap/download/setup.html](https://coastwatch.pfeg.noaa.gov/erddap/download/setup.html)
7. Pick an existing Converter method that is similar to what you want to make (e.g., gov.noaa.pfel.erddap.Erddap.doConvertOceanicAtmosphericAcronyms() ), find that code and make a duplicate of it right below it, with the new name doConvertColorBar.  (Also make a copy of the line of code that call that method, so it calls your code.)
8. Modify your method to do what you want it to do (without modifying any other code).
9. Test by hand by running ERDDAP on your computer.
10. Make a copy of Erddap.testConvertInterpolate() and write a unit test for your code. 

That looks like a lot, but it's just detailed instructions. This is a very do-able project.

I hope that makes sense. Email if/when you get stuck, but also at step 5 so we can talk about and agree on your plan.
Thanks again for giving this a try.

P.S. Can I also ask: how did you find this project to work on?  Were you already following ERDDAP or are the issues posted elsewhere?

**Comment by GAyan17** [comment]
Thank you @BobSimons  I'll surely follow these steps.

**Comment by BobSimons** [comment]
@GAyan17, are you still planning to work on this?

**Comment by GAyan17** [comment]
Yes, I am sorry I was busy with some other work so didn't get the chance to dive deeper into it, but I will surely try to work on it in this summer.

**Comment by Harish-2003** [comment]
hii @BobSimons i would like to work upon this issue and this info https://github.com/ERDDAP/erddap/issues/44#issuecomment-780896867 seems to be very much useful i will go ahead and make it implemented in my system will connect if i got any doubts this issue is still open right?

**Comment by Harish-2003** [comment]
i am having windows version is it okay if i go ahead with this?

**Comment by ChrisJohnNOAA** [comment]
@Harish-2003 Bob has retired, he's no longer active on the project. Yes this issue is still open.

One thing I'd like to add that was not covered above in the thread is to include tests of the new feature. The project is using JUnit and you can find many examples of tests in the src/test directory.

**Comment by Harish-2003** [comment]
https://erddap.github.io/docs/server-admin/deploy-install#programmersGuide i was currently following with this setup i think this one is not old? or do we have any better way to setup?

**Comment by ChrisJohnNOAA** [comment]
> https://erddap.github.io/docs/server-admin/deploy-install#programmersGuide i was currently following with this setup i think this one is not old? or do we have any better way to setup?
> 

You can use Jetty to run the server for development. There's some directions here: https://github.com/ERDDAP/erddap/tree/main/development/jetty

**Comment by Harish-2003** [comment]
![Image](https://github.com/user-attachments/assets/62f6688c-6e26-4837-ae6c-1be39e6609de)
i was getting this in my cmd how can i solve this one

**Comment by Harish-2003** [comment]
![Image](https://github.com/user-attachments/assets/b48f5016-20c4-4e94-8bb9-8e048dfd0b4e)
while trying to run this command `mvn jetty:run -DerddapContentDirectory=development/jetty/config` i got the above build error

**Comment by ChrisJohnNOAA** [comment]
> ![Image](https://github.com/user-attachments/assets/b48f5016-20c4-4e94-8bb9-8e048dfd0b4e) while trying to run this command `mvn jetty:run -DerddapContentDirectory=development/jetty/config` i got the above build error [@ChrisJohnNOAA](https://github.com/ChrisJohnNOAA) can you suggest me further steps that i can do to resolve this error Thanks

This likely means the zip file was corrupted during download. You may need to remove the bad file from your m2 cache. You may find some information on this page useful for getting the m2 directory:
https://confluence.atlassian.com/bamkb/deleting-maven-m2-and-general-cache-locations-used-by-bamboo-1271268165.html

The download plugin we use stores files in:
{pathToM2}/.m2/repository/.cache/download-maven-plugin

As for the other error, reading about MAVEN_OPTS might be useful for you: https://maven.apache.org/configure.html
Depending on your exact configuration it is also possible you don't need those specific options set.

**Comment by Harish-2003** [comment]
![Image](https://github.com/user-attachments/assets/f7be9d35-87bc-4a8d-819e-2b250554aaa7)
all the errors has been resolved but it got stuck here ? what am i supposed to do ? do i need to wait

**Comment by Harish-2003** [comment]
![Image](https://github.com/user-attachments/assets/b646617c-0de5-4033-8681-155819a90e31)
@ChrisJohnNOAA

**Comment by Harish-2003** [comment]
![Image](https://github.com/user-attachments/assets/ded89fca-dddf-4db5-93de-34eebf53b960)
succesfully running thank you @ChrisJohnNOAA

**Comment by Harish-2003** [comment]
as i am using windows version i have used this 
`set MAVEN_OPTS=-Xms4096m -Xmx4096m && mvn jetty:run -DerddapContentDirectory=development/jetty/config`
but in the readme it's metioned as 
`MAVEN_OPTS=-Xms4096m -Xmx4096m && mvn jetty:run -DerddapContentDirectory=development/jetty/config`
should i go ahead and make a pr for that one too?

**Comment by ChrisJohnNOAA** [comment]
I'd recommend looking at how one of the existing conversion helper pages is handled, for example the time conversion (doConver Time) page is here:
https://github.com/ERDDAP/erddap/blob/main/WEB-INF/classes/gov/noaa/pfel/erddap/Erddap.java#L21815

I'd recommend creating a new page, probably under the path convert/colorbar.html which would be where you will implement the color bar converter.

**Comment by Harish-2003** [comment]
so you are suggesting that i need to implement an function with some naming like colorbar_converter in the file  [https://github.com/ERDDAP/erddap/blob/main/WEB-INF/classes/gov/noaa/pfel/erddap/Erddap.java](https://github.com/ERDDAP/erddap/blob/main/WEB-INF/classes/gov/noaa/pfel/erddap/Erddap.java#L21815) 
and later linking it to the colorbar.html web page?
am i right?

**Comment by Harish-2003** [comment]
hi @ChrisJohnNOAA can you tell me where can i able to find these .html files ?

**Comment by Harish-2003** [comment]
hi @ChrisJohnNOAA can you tell me the path for the .html files 
where can i find the code for this particular file ?
https://coastwatch.pfeg.noaa.gov/erddap/convert/index.html

**Comment by Harish-2003** [comment]
There is code in ERDDAP to convert a data value to a color. and also  bob mentioned there is the code for this i need to implement to reverse do you know where the code is at

**Comment by Harish-2003** [comment]
actually i tried developing some code in normal cmd i have been entering the values like this and output looks like this 

![Image](https://github.com/user-attachments/assets/7f035b60-628a-4ffd-9098-4ecaf6fbf15c)
if its good then i would like to add it to the main file and map it accordingly
@ChrisJohnNOAA 
sorry for delay got stuck in some work

**Comment by Harish-2003** [comment]
sorry i have been busy this 3 weeks with some of my personal work and travel stuff will try to contribute my work within few days and let u know @ChrisJohnNOAA

**Comment by Harish-2003** [comment]
![Image](https://github.com/user-attachments/assets/e073066a-a18b-4e3e-b37d-99e1798519ed)
i can able to add the link but i can't able to see my page in the html itself?

![Image](https://github.com/user-attachments/assets/c810c549-8fb1-4d5c-936d-1b3b875631b9)
i have added the code but i think i am stuck with the linking thing can you help me out with this?

**Comment by ChrisJohnNOAA** [comment]
Without seeing your code it's difficult to know.

My best guess is It might be helpful for you to look at [doConvert in Erddap.java](https://github.com/ERDDAP/erddap/blob/main/WEB-INF/classes/gov/noaa/pfel/erddap/Erddap.java#L19212). If you didn't already, you likely need to add code for "Color.html" in there.

**Comment by Harish-2003** [comment]
[CompoundColorMap:getColor](https://github.com/ERDDAP/erddap/blob/main/WEB-INF/classes/gov/noaa/pfel/coastwatch/sgt/CompoundColorMap.java#L559)
i am writing the same logic that is present in the above file as of now is it okay to do so or do you suggest any other proposals?
@ChrisJohnNOAA

**Comment by Harish-2003** [comment]
and can you suggest me any project in meanwhile so that i will try to prepare an doc for gsoc2025 and share the same with u @ChrisJohnNOAA

**Comment by Harish-2003** [comment]
![Image](https://github.com/user-attachments/assets/366e2f5a-3bd0-45dd-a66b-7aab239154df)
just a trail will try to add color thing in the meanwhile can you suggest me any enhancments that i need to do for better conversions

**Comment by Harish-2003** [comment]
Can u provide me ur email id so that I will share u my proposal document for the same
@ChrisJohnNOAA as Tommorow is the deadline for it

**Comment by ChrisJohnNOAA** [comment]
> Can u provide me ur email id so that I will share u my proposal document for the same [@ChrisJohnNOAA](https://github.com/ChrisJohnNOAA) as Tommorow is the deadline for it

My e-mail address is chris.john@noaa.gov

You should be submitting an application through the Google Summer of Code site: https://summerofcode.withgoogle.com/

**Comment by Harish-2003** [comment]
![Image](https://github.com/user-attachments/assets/9d8f9e40-aee4-4ab3-9c7e-433114f0e0f7)
and if possible can you give me more info about color palette 
https://erddap.github.io/docs/server-admin/additional-information?_highlight=color#palettes u mean this?

**Comment by ChrisJohnNOAA** [comment]
> ![Image](https://github.com/user-attachments/assets/9d8f9e40-aee4-4ab3-9c7e-433114f0e0f7) and if possible can you give me more info about color palette https://erddap.github.io/docs/server-admin/additional-information?_highlight=color#palettes u mean this?

Yeah, you should provide a dropdown selector like this one on the .graph page:

![Image](https://github.com/user-attachments/assets/62bda013-09d9-49df-b486-b1721d24fc00)

I believe that is controlled by this section of [EDDGrid:respondToGraphQuery](https://github.com/ERDDAP/erddap/blob/main/WEB-INF/classes/gov/noaa/pfel/erddap/dataset/EDDGrid.java#L4605-L4634)

**Comment by Harish-2003** [comment]
![Image](https://github.com/user-attachments/assets/34906d60-7f4b-4228-ab79-b9af40eae5e6)
just giving a brief will do more modifications

**Comment by Harish-2003** [comment]
![Image](https://github.com/user-attachments/assets/c38aedfa-e62b-4824-a4b2-13bd347377f0)
brief comparision

![Image](https://github.com/user-attachments/assets/95d88ef6-692d-4c04-a4aa-c5a3a1f3e809)

![Image](https://github.com/user-attachments/assets/6bdb7630-3d25-4607-8ede-7027c869f976)

**Comment by ChrisJohnNOAA** [comment]
That's great progress.

Can you either make the result text black, or match the color?

Can you add the color bar based on the selected palette to the screen so a user can see what the full bar looks like?

**Comment by Harish-2003** [comment]
> That's great progress.
Thank u it means a lot.
> Can you either make the result text black, or match the color?
Yaa sure ,actually I haven't looked into it.i will make changes 
> Can you add the color bar based on the selected palette to the screen so a user can see what the full bar looks like?
Yepp will add and will share screenshot with the suggested changes

**Comment by Harish-2003** [comment]
![Image](https://github.com/user-attachments/assets/111f9abf-0072-417d-baef-e84932ace9bf)
i tried adding colors but when the color is white the values are also turing white so i prefered putting them in black 
and regarding the color bar i added this one as of now will try to add the colorbar like the one that is present in the 
[Graph web page](https://coastwatch.pfeg.noaa.gov/erddap/griddap/erdMWsstdmday_LonPM180.graph?sst%5B(2025-03-16T12:00:00Z)%5D%5B(0.0)%5D%5B(22.0):(51.0)%5D%5B(-155.0):(-105.0)%5D&.draw=surface&.vars=longitude%7Clatitude%7Csst&.colorBar=BlackGreenWhite%7CC%7CLinear%7C64%7C1000%7C2&.bgColor=0xffccccff
)
anything else you want me to do? or should i go ahead and make a pr?

**Comment by ChrisJohnNOAA** [comment]
Sounds great. A PR would be good. @Harish-2003

**Comment by Harish-2003** [comment]
Hi @ChrisJohnNOAA ,
i have a question , i just wanted to add those edstatic.messages to the colorbarconverter that we have added but i can't able  see the .xml files in my laptop is there anything i need to do  to see them .Even though i compiled it i can't able to find them i tried checking the messages.xml file but i couldn't able to find it

**Comment by ChrisJohnNOAA** [comment]
> Hi [@ChrisJohnNOAA](https://github.com/ChrisJohnNOAA) , i have a question , i just wanted to add those edstatic.messages to the colorbarconverter that we have added but i can't able  see the .xml files in my laptop is there anything i need to do  to see them .Even though i compiled it i can't able to find them i tried checking the messages.xml file but i couldn't able to find it

Hi @Harish-2003 
The files you need are [EDMessages.java](https://github.com/ERDDAP/erddap/blob/main/WEB-INF/classes/gov/noaa/pfel/erddap/util/EDMessages.java) and [messages.xml](https://github.com/ERDDAP/erddap/blob/main/src/main/resources/gov/noaa/pfel/erddap/util/messages.xml).

You may need to add the messages to all of the [translated files](https://github.com/ERDDAP/erddap/tree/main/src/main/resources/gov/noaa/pfel/erddap/util/translatedMessages). There are instructions for [running the translation script](https://github.com/ERDDAP/erddap/tree/main/translation), but if it gives you any problems, let me know.
